src/kernel/emulation/linux/misc/sysctl_machdep.c

```diff
@@ -51,3 +89,131 @@ sysctl_handler(handle_brand_string)
 	return 0;
 }
 
+sysctl_handler(handle_max_basic)
+{
+	setup(0);
+
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+
+
+
+	copyout_int(eax, (char*)old,oldlen);
+
+	return 0;
+}
+
+sysctl_handler(handle_family)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 7;
+    eax &= 15;
+
+    copyout_string(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_model)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 3;
+    eax &= 15;
+
+    copyout_int(eax,(char*)old, oldlen);
+
+    return 0;
+}
+
+sysctl_handler(handle_stepping)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax &= 15;
+
+    copyout_int(eax,(char*)old, oldlen);
+
+    return 0;
+}
+
+sysctl_handler(handle_brand_string)
+{
+    setup(0x80000000);
+
+    __cpuid(level,eax,eabx, eacx, eadx);
+
+    if(eax < 0x80000004) // the information is not implemented
+        return 2;
+
+
+
+    union
+    {
+        unsigned int brand[12];
+        char name[49];
+    } v;
+
+		for (int i = 1; i < 4; i++)
+		{
+				eax = 0x80000000+i;
+				__cpuid(level, eax, ebx, ecx, edx);
+				v.brand[0x0+(i-1)*4] = eax;
+				v.brand[0x1+(i-1)*4] = ebx;
+				v.brand[0x2+(i-1)*4] = ecx;
+				v.brand[0x3+(i-1)*4] = edx;
+		}
+
+    v.name[48] = 0;
+    copyout_string(v.name, (char*) old, oldlen);
+
+    return 0;
+}
+sysctl_handler(handle_features)
+{
+
+    if(old != NULL)
```

Comment:
Like I mentioned before, `sysctl` can be called with a `NULL` value for `old` in order to determine the length of the output string. This is partially correct (it avoids segfaults due to `NULL` values for `old`), but you don't have a branch for actually determining and outputting the length of the output string.

You could do this by moving lines 185 through 191 outside this `if` and add an `else` to the condition that is executed when `old` is `NULL`. In that branch, you just have to add up the lengths of the features that are set (plus the spaces in between).

That would look something like this:
```c
 else {
        int len = 0;
        for (int i = 0; i < 32; i++)
        {

            if(i == 10 || i == 20)
                continue;
            if(edx>>i&1) {
                if (len != 0)
                    len++;
                len += __simple_strlen(features[i]);
            }
        }
        if (oldlen)
            *oldlen = len;
    }
```